Added GitHub API collaborator matching between users and github_repos

ProbablyFaiz opened the pull request:
~NOTE: Do **not** merge this before merging the changes made in #115, which restructures job functionality. Then, those changes can be rebased into this branch for a much less painful process than if it were rebased the other way around.~

This was supposed to be a simple job implementation but ended up requiring a bit more due to the way Rails handles join tables. This PR does the following:

* Implements GitHub repo to user matching by asking the GitHub API for a list of direct collaborators to a repository. Previously, repositories were displayed on a roster_student's page via finding the student/user's username in the repository name. Now, substring and API matching is used, and the method of how the repo was matched is displayed:
![Screen Shot 2020-01-04 at 5 07 05 PM](https://user-images.githubusercontent.com/34611522/71773619-b1610b80-2f14-11ea-9e89-bc26bc6ffacc.png)
* In order to store how a user and repository were matched, a join table between users and repos called RepoContributors is explicitly created with extra columns indicating the matching methods used.
* The Refresh GitHub Repos job now does user-repo matching using the new Repo Collaborator job. We should talk about this as the performance hit associated could be significant due to how many API requests the Repo Contributor job makes.
* Fixed various bugs relating to GithubRepo deletion and Octokit errors.

This PR closes #107

pconrad commented:
@ProbablyFaiz do you mean that #115 should be merged first, and then this branch rebased?

I'm assuming that can be our work at 5:30pm

pconrad commented on app/jobs/repo_collaborators_job.rb:
```diff
@@ -0,0 +1,61 @@
+      repo_collaborators = []
+    end
+    students.each do |student|
+      unless student.user.nil? or student.user.username.nil?
```
Wait, really!?    

So, looping over all the students for each repo makes this O(s * r) where s is | students | and r is | repos|

I wonder if we could do better by simply parsing the repo name breaking it up where there are underscores.   The convention is that since hyphens are allowed in github usernames we require that underscores be used to separate the lab name from the two github usernames.

I'm seeing that this convention is pretty consistent across Diba, Richert, and my courses.    So I think we can depend on it.

Also: don't hate me---but I'm starting to wonder whether we need the substring convention at all now that we have the more meaningful and reliable collaborator API listing.  That's really the ground truth.

But, your table might not have been a waste of time.  It may be helpful to have the information about what "type" of collaborator each user is (admin, read, write, etc.).    We might want that in place of knowing whether the match was found by the API or by the substring method.

ProbablyFaiz replied:
Substring matching removed in 4d38df8

pconrad commented:
Let's remove the substring matching, and replace the columns in that database table with a column that has the collaborator `permission_level`.   

For now, we can just use `read`, `write` and `admin`, ignoring the existence of `triage` and `maintain` given what we learned (i.e. that those are not yet supported in the GitHub API).

Later on, if support for those is added, we can fix that. 

So, we'll display the collaborator level.  It's typically the case that for most students assignments, we want one or read, write, or admin anyway, and not triage or maintain.  For those very small number of use cases, we'll just have to have the teaching staff check manually.

ProbablyFaiz commented:
Reworked the show student repo table as discussed. Picture below:

<img width="1136" alt="Screen Shot 2020-01-07 at 11 21 19 PM" src="https://user-images.githubusercontent.com/34611522/71958509-55f97c80-31a5-11ea-8380-77d10c9fb6d1.png">

Additionally, I have removed substring matching and the RepoContributor columns that tell how a repo was matched to a user. We should discuss the substring matching one more time before merging this, however, as I believe I have discovered some edge cases where it will be necessary because of the collaborator API call's quirks. 

The RepoCollaborator job is now a separate job again but is no longer triggered automatically by RefreshGithubReposJob. It must be called separately. #123 has not yet been implemented, so any instructor can call this job but it should be okay for the time being.

scottpchow23 commented on app/jobs/refresh_github_repos_job.rb:
```diff
@@ -10,14 +10,15 @@ def perform(course_id)
           num_created += create_repo_record(github_repo, course, students)
         end
         summary = num_created.to_s + " repos created, " + (org_repos.size - num_created).to_s + " repos refreshed."
+        # RepoCollaboratorsJob.perform_async(course_id)
```
any reason why we're leaving commented code here?

ProbablyFaiz replied:
From one of my above comments:
>The RepoCollaborator job is now a separate job again but is no longer triggered automatically by RefreshGithubReposJob.

It may at some point be desired functionality to update all collaborators from this job as a kind of hard refresh of all GitHub data. For now, we aren't doing that but I left that line commented just as a placeholder if we ever do. Let me know if I should take it out.

scottpchow23 replied:
No that's fine then; just needed to remember why we want to leave it there. Lgtm.
